File: app/auth/modules/basic_auth.py
from __future__ import annotations

import base64
import os
from typing import Any

import httpx
from redis.asyncio import Redis

from app.auth.base import AuthContext, AuthResult, _invalidate_on_from_config


class BasicAuth:
    """HTTP Basic authentication from two environment variables."""

    name = "basic_auth"

    def __init__(self, encoded: str, invalidate_on: set[int]) -> None:
        self._encoded = encoded
        self._invalidate_on = invalidate_on

    @classmethod
    def from_config(
        cls, config: dict[str, Any], *, redis: Redis, http: httpx.AsyncClient
    ) -> BasicAuth:
        username_env = config["username_env"]
        password_env = config["password_env"]
        username = os.environ.get(username_env)
        if not username:
            raise ValueError(f"BasicAuth: env var {username_env!r} is not set or empty.")
        password = os.environ.get(password_env)
        if not password:
            raise ValueError(f"BasicAuth: env var {password_env!r} is not set or empty.")
        encoded = base64.b64encode(f"{username}:{password}".encode()).decode()
        return cls(encoded=encoded, invalidate_on=_invalidate_on_from_config(config))

    async def apply(self, ctx: AuthContext) -> AuthResult:
        return AuthResult(headers={"Authorization": f"Basic {self._encoded}"})

    async def invalidate(self) -> None:
        pass

    def is_rejection(self, status_code: int, body: dict | None) -> bool:
        return status_code in self._invalidate_on

File: app/ts/src/auth/modules/basic_auth.ts
import type { AuthContext, AuthResult } from "../../core/types.ts";
import type { AuthModule, AuthDeps } from "../types.ts";
import { invalidateOnFromConfig } from "../types.ts";

export class BasicAuth implements AuthModule {
  static readonly typeName = "basic_auth";

  // Encoded once at construction — avoids re-encoding on every request.
  private readonly encoded: string;
  private readonly invalidateOn: Set<number>;

  private constructor(encoded: string, invalidateOn: Set<number>) {
    this.encoded = encoded;
    this.invalidateOn = invalidateOn;
  }

  static fromConfig(
    config: Record<string, unknown>,
    deps: AuthDeps,
  ): BasicAuth {
    if (typeof config["username_env"] !== "string") {
      throw new Error("BasicAuth: 'username_env' config field is required.");
    }
    if (typeof config["password_env"] !== "string") {
      throw new Error("BasicAuth: 'password_env' config field is required.");
    }

    const userEnv = config["username_env"];
    const passEnv = config["password_env"];

    const user = deps.env[userEnv];
    if (typeof user !== "string" || !user) {
      throw new Error(`BasicAuth: env var '${userEnv}' is not set or empty.`);
    }

    const pass = deps.env[passEnv];
    if (typeof pass !== "string" || !pass) {
      throw new Error(`BasicAuth: env var '${passEnv}' is not set or empty.`);
    }

    // btoa is available in Node 22+ and Cloudflare Workers globally.
    const encoded = btoa(`${user}:${pass}`);
    const invalidateOn = invalidateOnFromConfig(config, [401]);

    return new BasicAuth(encoded, invalidateOn);
  }

  async apply(_ctx: AuthContext): Promise<AuthResult> {
    return { headers: { Authorization: `Basic ${this.encoded}` } };
  }

  async invalidate(): Promise<void> {}

  isRejection(statusCode: number, _body: unknown): boolean {
    return this.invalidateOn.has(statusCode);
  }
}
